flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/over/NonTimeRangeUnboundedPrecedingFunction.java

```diff
@@ -0,0 +1,596 @@
+            if (sortKeyEqualiser.equals(curSortKey, inputSortKey)) {
+                setAccumulatorOfPrevRow(sortedList, i - 1);
+                // Accumulate all ids except id which needs to be removed
+                int removeIndex = reAccumulateIdsAndGetRemoveIndex(ids, input);
+                if (removeIndex == -1) {
+                    // TODO: can this ever happen? may be during/after state cleaning?
+                    LOG.info("Could not find id to remove");
+                    return;
+                }
+                emitUpdatesForIds(
+                        ids,
+                        removeIndex,
+                        accMapState.get(curSortKey),
+                        currFunction.getAccumulators(),
+                        RowKind.DELETE,
+                        input,
+                        out);
+                Long deletedId = ids.remove(removeIndex);
+                i = removeIdFromSortedList(sortedList, i, ids, curSortKey);
+                removeFromState(sortedList, deletedId, curSortKey);
+                break;
+            }
+        }
+
+        processRemainingElements(sortedList, i, out);
+    }
+
+    /**
+     * Helper method to re-accumulate the aggregated value for all ids without the id that will be
+     * removed.
+     *
+     * @param ids
+     * @param input
+     * @return the index position of the id that should be removed
+     * @throws Exception
+     */
+    private int reAccumulateIdsAndGetRemoveIndex(List<Long> ids, RowData input) throws Exception {
+        int removeIndex = -1;
+        for (int j = 0; j < ids.size(); j++) {
+            RowData curValue = valueMapState.get(ids.get(j));
+            if (valueEqualiser.equals(curValue, input)) {
+                removeIndex = j;
+            } else {
+                currFunction.accumulate(curValue);
+            }
+        }
+        return removeIndex;
+    }
+
+    /**
+     * Remove the ith element from the sortedList if the list of ids is empty and return the ith
+     * location OR update the sorted list with the list of ids and return the next location of the
+     * sortedList.
+     *
+     * @param sortedList
+     * @param idx
+     * @param ids
+     * @param curSortKey
+     * @return
+     */
+    private int removeIdFromSortedList(
+            List<Tuple2<RowData, List<Long>>> sortedList,
+            int idx,
+            List<Long> ids,
+            RowData curSortKey) {
+        if (ids.isEmpty()) {
+            sortedList.remove(idx);
+        } else {
+            // Update sortedList without the removed id
+            sortedList.set(idx, new Tuple2<RowData, List<Long>>(curSortKey, ids));
+            idx++;
+        }
+        return idx;
+    }
+
+    private void removeFromState(
+            List<Tuple2<RowData, List<Long>>> sortedList, Long id, RowData sortKey)
+            throws Exception {
+        valueMapState.remove(id);
+        accMapState.put(sortKey, currFunction.getAccumulators());
```

Comment:
if this was the last row with the same sort key, we have to delete the accumulator from the map.

Reply:
Fixed

Reply:
Also added a test in `NonTimeRangeUnboundedPrecedingFunctionTest` to validate state content/size.